fdp/fdp_settings.md
# FDP Information System Settings

## Overview

The Full Disclosure Project (FDP) information system can be customized through the use of configuration settings. For most projects, all relevant settings can be specified in environment variables that are accessible by the web application or in configuration files stored in the *../conf/* folder. In some cases, projects can choose to specify additional settings in the *fdp/settings.py* file.

## Django Settings

Many of the available settings for the FDP system are defined and handled through the Django web framework upon which the system is built.

For a full reference of Django settings, see: <https://docs.djangoproject.com/en/3.2/ref/settings/>

There are additional settings available for the FDP system that are defined by the Python packages which are integrated into it. These packages vary based on the hosting environment selected.

## Hosting Environment

The settings for the FDP system have been preconfigured for selected hosting environments, and can be managed in the *fdp/settings.py* file. 

To configure for any hosting environment, add a comment in front of all `import` statements.

To configure for hosting in a *local development* environment, remove the comment from in front of:

    from .local_settings import *

To configure for hosting in a *Microsoft Azure* environment, remove the comment from in front of:

    from .azure_settings import *

Once a hosting environment is configured, specify its relevant settings using the below table.

## Microsoft Azure Active Directory

When hosting in a Microsoft Azure environment is configured as described above, user authentication through an Azure Active Directory can be enabled by specifying the relevant Azure Key Vault secrets and Azure App Service application settings.

Additionally, to enforce user authentication through *only Azure Active Directory*, remove the comment from in front of:

    from .azure_backend_only_settings import *

## Logging

A preconfigured logging mechanism is available in the FDP system, and can be turned on in the *fdp/settings.py* file.

To enable file logging, remove the comment in front of the following line, and specify the file path:

    FDP_ERR_LOGGING['handlers']['file']['filename'] = '/home/site/wwwroot/debug.log'

Then, to allow the assignment to the appropriate Django setting, remove the comment in front of the following line:

    LOGGING = FDP_ERR_LOGGING

## User-uploaded files

Files can be uploaded by users into the FDP system, and linked to instances of both the `Attachment` and `PersonPhoto` models. Though their specifics are different, both have a maximum number of bytes for the file size and vet uploaded files against a list of supported file types. 

Changing the default maximum number of file size bytes for either models will require generating corresponding database migrations through Django's `makemigrations` and `migrate` commands. 

To change the default maximum number of file size bytes for the `Attachment` model, specify the number of bytes:

    FDP_MAX_ATTACHMENT_FILE_BYTES = 104857600

To change the default maximum number of file size bytes for the `PersonPhoto` model, specify the number of bytes:

    FDP_MAX_PERSON_PHOTO_FILE_BYTES = 2097152

Supported file types are defined by a list of tuples, with the first item being a user-friendly description of the supported file type, and the second item being the expected file extension in lowercase:

    ..., ('Microsoft Word 97-2003', 'doc'), ('Microsoft Word 2007+', 'docx'), ...

To add to the list of supported file types for the `Attachment` model:

    FDP_SUPPORTED_ATTACHMENT_FILE_TYPES.append(
        ('User-friendly description for file type', 'ext')
    )

To overwrite the *entire* list of supported file types for the `Attachment` model:

    FDP_SUPPORTED_ATTACHMENT_FILE_TYPES = [
        ... ('User-friendly description for file type', 'ext'), ...
    ]

To add to the list of supported file types for the `PersonPhoto` model:

    FDP_SUPPORTED_PERSON_PHOTO_FILE_TYPES.append(
        ('User-friendly description for file type', 'ext')
    )

To overwrite the *entire* list of supported file types for the `PersonPhoto` model:

    FDP_SUPPORTED_PERSON_PHOTO_FILE_TYPES = [
        ... ('User-friendly description for file type', 'ext'), ...
    ]

## Federated login page

A federated login page can be enabled and the options that are displayed on it can be customized by configuring:  

    FEDERATED_LOGIN_OPTIONS = [
        {
            'label': 'Sign in with FDP',
            'url_pattern_name': 'two_factor:login',
            'url_pattern_args': [],
            'css': {'background-color': '#417690', 'color': '#FFF'},
            'css_hover': {'color': '#f5dd5d'}
        },
        {
            'label': 'Sign in Azure Active Directory',
            'url_pattern_name': 'social:begin',
            'url_pattern_args': ['inheritable.models.AbstractConfiguration.azure_active_directory_provider'],
            'css': {'background-color': '#417690', 'color': '#FFF'},
            'css_hover': {'color': '#f5dd5d'}
        }
    ]

If no options are listed, the federated login page will be skipped, and the user will be automatically redirected to the primary login page depending on the system's configuration.

## EULA

To enable the requirement that each user agree to an end-user license agreement (EULA) before they can access any secured views, set `FDP_EULA_SPLASH_ENABLE=True`. By default, this feature is disabled.

## Wholesale import tool

The default list of models that are in the allowlist, and so allowed for use through the wholesale import tool, can be modified by specifying a new list of model names: 

    FDP_WHOLESALE_MODELS_ALLOWLIST = ['Attachment', 'Content', 'ContentIdentifier', ...]

The default list of fields that are in the denylist, and so denied from use through the wholesale import tool, can be modified by specifying a new list of field names: 

    FDP_WHOLESALE_FIELDS_DENYLIST = ['is_archived', ...]

## Legacy officer search page

Version 6.0.0 introduces a new officer search page, which replaces the legacy search page. To re-enable the old legacy 
search use this setting:

    LEGACY_OFFICER_SEARCH_ENABLE = True

## FDP Settings for Azure

For hosting in Microsoft Azure, it is recommended that the most sensitive settings are stored in Microsoft Azure Key Vault as Secrets. Less sensitive settings can be stored in the Azure App Service as Application Settings. The following is one possible configuration:

| Required | Azure Key Vault Secret | Azure App Service Application Setting (environment variable) | Default | Details |
| --- | --- | --- | --- | --- |
|  |  | FDP_EXTERNAL_AUTHENTICATION | `None` | Additional authentication mechanism. Use `'aad'` for Azure Active Directory. |
| Required | | FDP_AZURE_KEY_VAULT_NAME |  | Name for Azure Key Vault. Must be defined to use Secrets in Azure Key Vault. |
| Required | FDP-SECRET-KEY | | |  Secret key for Django. |
| | | FDP_ALLOWED_HOST | From environment variable: `'WEBSITE_HOSTNAME'` | Host for Django to serve. |
| | | FDP_DATABASE_NAME | `'fdp'` | Name of database. |
| Required | FDP-DATABASE-USER | | | Username for database. |
| Required | FDP-DATABASE-PASSWORD | | | Password for database. |
| Required | | FDP_DATABASE_HOST | | Host for database access. |
| | | FDP_DATABASE_PORT | `'5432'` | Port for database access. |
| Required | FDP-QUERYSTRING-PASSWORD | | | URL-safe base64-encoded 32-byte key for querystring encryption. |
| Required | | FDP_AZURE_STATIC_STORAGE_ACCOUNT_NAME | | Name for Azure Storage account with static container. |
| Required | FDP-AZURE-STATIC-STORAGE-ACCOUNT-KEY *(specify access key value)* | FDP_AZURE_STATIC_STORAGE_ACCOUNT_KEY *(specify reference to secret)* | | Access key for Azure Storage account with static container. Specify both in Azure Key Vault (using access key value), and in Azure App Service Application Setting (using reference to secret). |
| Required | | FDP_AZURE_MEDIA_STORAGE_ACCOUNT_NAME | | Name for Azure Storage account with media container. |
| Required | FDP-AZURE-MEDIA-STORAGE-ACCOUNT-KEY *(specify access key value)* | FDP_AZURE_MEDIA_STORAGE_ACCOUNT_KEY *(specify reference to secret)* | | Access key for Azure Storage account with media container. Specify both in Azure Key Vault (using access key value), and in Azure App Service Application Setting (using reference to secret). |
| | | FDP_AZURE_STORAGE_ACCOUNT_SUFFIX | `'blob.core.windows.net'` | Suffix for Azure Storage account. |
| | | FDP_AZURE_STATIC_CONTAINER | `'static'` | Name for static files container. |
| | | FDP_AZURE_STATIC_EXPIRY | `None` | Seconds for static file expiration. |
| | | FDP_AZURE_MEDIA_CONTAINER | `'media'` | Name for media files container. |
| | | FDP_AZURE_MEDIA_EXPIRY | `20` | Seconds for media file expiration. |
| | FDP-SOCIAL-AUTH-AZUREAD-TENANT-OAUTH2-KEY | | | Azure Active Directory client ID. Only checked if `'aad'` was specified for Azure Active Directory. |
| | FDP-SOCIAL-AUTH-AZUREAD-TENANT-OAUTH2-TENANT-ID | | | Azure Active Directory tenant ID. Only checked if `'aad'` was specified for Azure Active Directory. |
| | FDP-SOCIAL-AUTH-AZUREAD-TENANT-OAUTH2-SECRET | | | Azure Active Directory client secret. Only checked if `'aad'` was specified for Azure Active Directory. |
| | | FDP_SOCIAL_AUTH_AZUREAD_TENANT_OAUTH2_WHITELISTED_DOMAINS | | Comma-separate list of Azure Active Directory domains. Only checked if `'aad'` was specified for Azure Active Directory. |
| | | FDP_FROM_EMAIL | `'webmaster@localhost'` | `FROM` email address for system. |
| | | FDP_EMAIL_HOST | `'localhost'` | Host used to send emails. |
| | FDP-EMAIL-HOST-USER | | | User for login to send emails. |
| | FDP-EMAIL-HOST-PASSWORD | | | Password for login to send emails. |
| | | FDP_EMAIL_PORT | `25` | Port to send emails. |
| | | FDP_RECAPTCHA_PUBLIC_KEY | | Public reCAPTCHA key. |
| | FDP-RECAPTCHA-PRIVATE-KEY | | | Private reCAPTCHA key. |

## FDP Settings for Local Development

For hosting in a local development environment, it is recommended that settings are stored in configuration files. The following is one possible configuration:

| Required | Configuration File (.conf) | Default | Details |
| --- | --- | --- | --- |
| | fdp_from_email | `'webmaster@localhost'` | `FROM` email address for system. | 
| | fdp_email_host | `'localhost'` | Host used to send emails. | 
| | fdp_email_host_user | | User for login to send emails. | 
| | fdp_em | | Password for login to send emails. | 
| | fdp_email_port | `25` | Port to send emails. | 
| | fdp_recaptcha_public_key | | Public reCAPTCHA key. | 
| | fdp_ca | | Private reCAPTCHA key. | 
| Required | fdp_sk | | Secret key for Django. | 
| | fdp_database_name | `'fdp'` | Name of database. | 
| | fdp_database_user | `'django_fdp'` | Username for database. | 
| Required | fdp_ps | | Password for database. | 
| | fdp_database_host | `'localhost'` | Host for database access. | 
| | fdp_database_port | `'5432'` | Port for database access. | 
| Required | fdp_qs | | URL-safe base64-encoded 32-byte key for querystring encryption. | 


CHANGELOG.md
# Changelog

All releases will be logged in this file.

## [6.0.2] - 2022-08-11

### Changed
- Minor typographical and user interface fixes

## [6.0.1] - 2022-08-09

### Changed
- Importer: Handle Microsoft Byte Order Mark on CSV import sheets

## [6.0.0] - 2022-08-03

### Added
- New Person search page with improved usability ('roundup' search)
- New Person search algorithm that better handles middle names, abbreviations, variation in spelling and 
  punctuation ('roundup' search)
- Newly designed bulk importer workflow pages with more specific error reporting and guidance
- Bulk importer columns documentation page
- Command line interface for bulk importer
- Command line data exporter tool (no Web UI yet)
- Custom text blocks on site footer and profile pages
- Help text for data fields on editing pages
- Optional default types and tags (available on installation)

### Changed
- Add Group field to Person Title records
- Rename "is inactive" field to "ended unknown date"
- Add "ended unknown date" to date fields on PersonIdentifier, PersonRelationship, PersonPayment, 
  GroupingRelationship, PersonTitle, PersonGrouping, ContentCase, and Incident; and add "at least since" field on 
  ContentCase, and Incident
- Fix relationship types not updating on Person and Group editing pages
- Add audit log when AAD attempted logins have incorrect domain (security)
- Fix carriage returns in description text not showing on detail pages
- Increase default max failed password attempts and decrease default lockout time
- Fix empty parens when rank dates set to all zeros
- Add missing 'host only' filter option in Advanced Admin listing pages
- Rename "Officer" to "Person" on homepage, and officer search pages

### Removed
- "Snapshot" feature on person profile page now disableable via settings.py setting
- Prior version of Officer search disabled by default, but reenablable via settings.py setting

### Security patches
- Upgrade Django to 3.2 LTS release from deprecated 3.1 release channel
- PyJWT 2.4.0
- sqlparse 0.4.2
- lxml 4.9.1

### Upgrading notes
The new 'roundup' search requires that the TrigramSimilarity extension be enabled in PostgreSQL, and adds a new 
"full_text" field. This change is handled in a migration file. There are several other new fields and a field rename.

To accomplish all of this database migrations must be run during upgrade:
```shell
python3 manage.py migrate
```
You may get 500 errors until you do this.

In addition, the "full_text" field must be populated during upgrade using a custom management command:

```shell
python3 manage.py reindex_search
```

This may take a while to run. Search will not function properly until this is done. Note that moving forward the 
full_text field automatically updates with any new or updated records, so it is unnecessary to run this command 
again in the future after the upgrade.

## [5.0.1] - 2022-03-07

### Changed

- Wholesale batch listing: fix incorrect sorting order
- Wholesale: fix backward-incompatibility with existing external ids table field use
- Changing pages: fix slow scroll animation when adding new items in inline formsets on changing pages

## [5.0.0] - 2022-02-22

### Added

  - Officer profile page: Add contact info to identification section
  - Edit records directly from officer profile page
  - Add new import tool ('wholesale' importer)
  - Add support center link to all pages for administrative users
  - Bulk data manipulation management commands: bulk_delete, export_external_ids, bulk_perms_host_only, 
    bulk_update_groups
  - Add functional testing framework for developers

### Changed

  - Attachments: add mp3 as valid upload file in default settings -- and expand list with other common formats
  - Change "as of" field on dates to "at least since" (see migration notes below)
  - Officer profile page:
    - Redesign identification section for improved readability
    - Group membership show officer's relation to group, link to group, and print "until unknown-end-date" when end 
      date is unknown
    - Payroll section, collapse out of the way by default, and print hyphens when values are empty
    - Redesign Associates section to: reflect subject object order, add dates of relationship, and link to 
      associates' profiles
    - Rename "Misconduct" to "Known incidents"
    - Reverse order of incidents
    - In incident details show persons' situation role and rename "other officers involved" to "others involved"
  - Upgrade Django to 3.1.14, and sqlparse to 0.4.2 (security releases)

### Removed

  - Officer profile: Bolding of various data points under certain conditions in identification section.

### Upgrading notes

Changes to the associates section of the officer profile page require that relationship types (i.e. predicates) be 
updated to include the preposition (e.g. 'of' 'with' 'for'). For example "Father" should be changed to "Father of" 
and "Sibling" should be changed to "Sibling with". These must be updated manually.

The new wholesale importer feature requires a database migration. Run:
```shell
python3 manage.py migrate
```
to apply these migrations to the database. You may get a 500 error until you do this.


## [4.0.0] - 2021-12-14

### Added
- Add federated login page
- Add license file to project root dir
- Add vagrant for local development / demo sandbox environment
- Add is law enforcement check-box for Groups - update search to only show marked groups

### Changed
- Make static and media files be located in separately define storage containers
- Make is law enforcement checkbox a select box and make it a required field

### Upgrading notes

**Groups**

The new is_law_enforcement field on Groups requires a database migration. Run:
```shell
python3 manage.py migrate
```
to apply these migrations to the database. *You will get a 500 error until you do this.*

The search *won't return any groups* that aren't set as law enforcement. To bulk set all existing groups as law 

fdp/settings.py
"""
Django settings for Full Disclosure project.

Generated by 'django-admin startproject' using Django 3+.

For more information on this file, see
https://docs.djangoproject.com/en/3.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.1/ref/settings/
"""


#: To configure hosting in a local development environment, remove the comment from the below import statement.
#: This should not be used in a production environment.
# from .configuration.local.local_settings import *
from .configuration.azure.azure_settings import *

#: To configure hosting in a Microsoft Azure environment, remove the comment from the below import statement.
# from .configuration.azure.azure_settings import *


#: To enforce user authentication only through the Azure Active Directory backend, remove the comment from the below
# import statement.
# from .configuration.azure.azure_only_settings import *


# Name of Python file containing class that defines person profile searches.
FDP_PERSON_PROFILE_SEARCH_FILE = 'def_person'
# Name of class inheriting from AbstractProfileSearch that defines person profile searches.
FDP_PERSON_PROFILE_SEARCH_CLASS = 'PersonProfileSearch'
# Name of Python file containing class that defines grouping profile searches.
FDP_GROUPING_PROFILE_SEARCH_FILE = 'def_grouping'
# Name of class inheriting from AbstractProfileSearch that defines grouping profile searches.
FDP_GROUPING_PROFILE_SEARCH_CLASS = 'GroupingProfileSearch'
# Name of Python file containing class that defines person changing searches.
FDP_PERSON_CHANGING_SEARCH_FILE = 'def_person'
# Name of class inheriting from AbstractChangingSearch that defines person changing searches.
FDP_PERSON_CHANGING_SEARCH_CLASS = 'PersonChangingSearch'
# Name of Python file containing class that defines incident changing searches.
FDP_INCIDENT_CHANGING_SEARCH_FILE = 'def_incident'
# Name of class inheriting from AbstractChangingSearch that defines person changing searches.
FDP_INCIDENT_CHANGING_SEARCH_CLASS = 'IncidentChangingSearch'
# Name of Python file containing class that defines content changing searches.
FDP_CONTENT_CHANGING_SEARCH_FILE = 'def_content'
# Name of class inheriting from AbstractChangingSearch that defines content changing searches.
FDP_CONTENT_CHANGING_SEARCH_CLASS = 'ContentChangingSearch'
# Name of Python file containing class that defines grouping changing searches.
FDP_CONTENT_GROUPING_SEARCH_FILE = 'def_grouping'
# Name of class inheriting from AbstractChangingSearch that defines content changing searches.
FDP_CONTENT_GROUPING_SEARCH_CLASS = 'GroupingChangingSearch'


#: To enable logging, remove the comments from the below assignments.
FDP_ERR_LOGGING['handlers']['file']['filename'] = BASE_DIR / 'debug.log'
LOGGING = FDP_ERR_LOGGING

# DEBUG SETTING
# Don't limit number of parameters in post/get requests
# https://docs.djangoproject.com/en/2.2/ref/settings/#data-upload-max-number-fields
# Resolves 400 error when submitting bulk change and clicking "Select all XYZ [objects]"
# https://stackoverflow.com/questions/55921865/django-admin-deleting-all-records-bad-request-400
# DATA_UPLOAD_MAX_NUMBER_FIELDS = None

SESSION_COOKIE_AGE = 60 * 60 * 24 * 14

